implement pipeline run

droot opened the pull request:
This PR implements new feature `config rendering`.  Hydration can be invoked using command: `kpt fn render`.

natasha41575 commented:
noticed this has test case for nested package, does it cover https://github.com/GoogleContainerTools/kpt/issues/1441? Checking to see if I still need to work on that or if I should assign that issue to you

droot commented:
@natasha41575 yes, I have added a basic test for nested UX so it covers #1441 to get some coverage on the code in this PR. Also note that the expected output is wrong because I am still debugging `set-label` function. 
We have to add more test cases. I do want to think if our current approach for adding tests is scalable (mainly from verbosity standpoint).

natasha41575 commented on internal/pipeline/executor.go:
```diff
@@ -22,174 +22,174 @@ import (
+
+	// SinkStdout means that the result of hydration will be
+	// written to stdout.
+	SinkStdout = "stdout"
```
any reason why SinkInplace is type SinkMode but the other two aren't?

droot replied:
Other two also need to have the Type. It was an oversight.
I took out the `sinkMode` functionality for now. `SinkMode` support will be added in a separate PR with `out-of-place` support.

Shell32-Natsu commented on internal/pipeline/testdata/app-with-db/.expected/config.yaml:
```diff
@@ -0,0 +1,2 @@
+exitCode: 0
+network: false
```
`network` is default to false

droot replied:
Got it. Took it out.

frankfarzan commented on internal/pipeline/executor.go:
```diff
@@ -36,159 +38,129 @@ type Executor struct {
+	p, err := pkg.New(path)
 	if err != nil {
-		return nil, fmt.Errorf("failed to find abs path %s %w", path, err)
+		return nil, fmt.Errorf("failed to read pkg %w", err)
```
general comment: choose the failure error messages shown to the user carefully. It's an important part of the UX. In this case, `pkg` is an internal variable name.

droot replied:
I agree. I have improved error messages to make it useful for the user for now, but I will scrub them again once we have the finalized the command UX and terminology.

I realize that errors are serving two users (implementor and the end-user), so striking the balance sometimes is getting difficult. I found this post from Rob Pike https://commandcenter.blogspot.com/2017/12/error-handling-in-upspin.html very useful. I want to explore if we can apply those lessons to kpt error handling.

frankfarzan commented on internal/pipeline/executor.go:
```diff
@@ -36,159 +38,129 @@ type Executor struct {
-}
+// subPackages returns paths to sub packages including current package's path.
+// It includes a package only if it has a Kptfile.
+func (pn *pkgNode) subPackages() ([]string, error) {
```
This functionality should be extracted out/abstracted and used across different features (I'm guessing this is re-implemented multiple times already):

pkg.New(path).SubPackages() -> []Pkg

I don't think there's any reason for hydration logic to do any low-level file operations.

droot replied:
Good point. I moved this to `pkg` abstraction.

frankfarzan commented on internal/pipeline/executor.go:
```diff
@@ -36,159 +38,129 @@ type Executor struct {
-	// TODO(droot): figure out how to avoid reading sub packages here ?
-	pkgReader := &kio.LocalPackageReader{PackagePath: p.Path(), MatchFilesGlob: kio.MatchAll}
+// localResources returns resources that belong to this package excluding the subpackage resources.
+func (pn *pkgNode) localResources(includeMetaResources bool) (resources []*yaml.RNode, err error) {
```
Same here:

pkg.New(path).LocalResources()

droot replied:
Good pt. Done.

frankfarzan commented on internal/pipeline/executor.go:
```diff
@@ -207,78 +179,84 @@ func (s hydrationState) String() string {
-			err = fmt.Errorf("invalid pkg state %v detected", state)
+			resources = currPkg.resources
+			return resources, err
+		case Dry:
```
why remove the default? It's more defensive in case you the enum changes in the future

droot replied:
Added `default`.

frankfarzan commented on internal/pipeline/executor.go:
```diff
@@ -207,78 +179,84 @@ func (s hydrationState) String() string {
+			resources = currPkg.resources
+			return resources, err
+		case Dry:
+			// Is this a dead path ?
```
?

when would the state ever be Dry?

droot replied:
When package are initialized, they are in `Dry` state, but the way our recursive hydration code is implemented, we never encounter the `Dry` state so if we come here, that represent `internal error`.

frankfarzan commented on internal/pipeline/executor.go:
```diff
@@ -288,25 +266,42 @@ func hydrate(p *pkg, hctx *hydrationContext) (resources []*yaml.RNode, err error
+	return resources, nil
+}
+
+// filterMetaResources filters kpt metadata files such as pipeline,
```
pipeline is no longer a separate resource?
what's permissions?

droot replied:
fixed the comment.

frankfarzan commented on internal/pipeline/executor.go:
```diff
@@ -288,25 +266,42 @@ func hydrate(p *pkg, hctx *hydrationContext) (resources []*yaml.RNode, err error
+// filterMetaResources filters kpt metadata files such as pipeline,
 // Kptfile, permissions etc.
-func filterMetaData(resources []*yaml.RNode) []*yaml.RNode {
+func filterMetaResources(resources []*yaml.RNode) []*yaml.RNode {
```
This should filter out function config specified in`pipeline.config` right? Please have a test for all the cases here.

frankfarzan commented on internal/pipeline/executor.go:
```diff
@@ -207,78 +179,84 @@ func (s hydrationState) String() string {
+			// path annotation in each resource points to path relative to that package.
+			// But the resources are written to the file system at the root package level, so
+			// the path annotation in each resources needs to be adjusted to be relative to the rootPkg.
+			relPath, err := filepath.Rel(hctx.root.Path(), s)
```
This is the kind of low-level operation that needs to be done once and reused instead of re-implemented multiple times.

Also how do you handle new resources generated in the package after executing the pipeline? Please compare to to how 'kpt fn run' works today.

droot replied:
I tweaked this to handle generated resources and also added a basic testcase for generator in a nested package. Also changed the order, it is not being done as a post-hydration step at the pkg level. I can potentially move it to a helper method now.

frankfarzan commented on internal/pipeline/executor.go:
```diff
@@ -207,78 +179,84 @@ func (s hydrationState) String() string {
-			transitiveResources, err = hydrate(depPkg, hctx)
+			transitiveResources, err = hydrate(subPkg, hctx)
 			if err != nil {
 				err = fmt.Errorf("failed to hydrated sub pkg %s %w", s, err)
```
ditto on error messages being part of the UX: does "hydrated" make sense to the end-user?

droot replied:
I am using `running pipeline` instead of `hydrated` for now and will scrub once again when we settle on the terminology.

frankfarzan commented on internal/pipeline/executor.go:
```diff
@@ -207,78 +179,84 @@ func (s hydrationState) String() string {
-	pkgWriter := &kio.LocalPackageWriter{PackagePath: p.Path()}
 
-	filters, err := fnFilters(p.Pipeline(), p.Path())
+	pl, err := currPkg.pkg.Pipeline()
```
assuming the following lines work correctly if pl is nil or empty

frankfarzan commented on internal/pipeline/testdata/app-with-db/configs/Kptfile:
```diff
@@ -0,0 +1,19 @@
+    configMap:
+      namespace: staging
+  - image: gcr.io/kpt-fn/set-label:unstable
+    config:
```
this should be a configMap for brevity

droot replied:
@Shell32-Natsu is addressing this, will use the `configMap` version once the PR is in.

Shell32-Natsu replied:
We have already updated to support ConfigMap. @mengqiy Have we released them?

mengqiy replied:
Yes, we have released them. Tag `unstable` is always from HEAD.

droot replied:
Thanks guys! Using `configMap` now, just verified with the latest `set-label` and it works.

frankfarzan commented on internal/pipeline/testdata/app-with-db/configs/db/Kptfile:
```diff
@@ -0,0 +1,19 @@
+kind: Kptfile
+metadata:
+  name: db
+pipeline:
```
you want to test:

- a generator function in subpackage `db`
- a generator function in package `app`
- pipeline that uses `configs` to exercise filtering meta resources.
- validation functions: shouldn't be allowed to mutate resources
- failure cases

droot replied:
This is not resolved yet. I will be addressing this soon and capture that in a separate commit.

droot replied:
Added testcases for filtering meta resources test case and generator in the subpackage `db` . Rest of the test cases I will be adding in the separate PR.

frankfarzan commented on internal/pkg/pkg.go:
```diff
@@ -77,23 +80,57 @@ func New(path string) (*Pkg, error) {
-		// TODO
-		// p.kptfile = ...
-		p.kptfileLoaded = true
+func (p *Pkg) Kptfile() (*kptfilev1alpha2.KptFile, error) {
```
Is this used in the hydration code?

droot replied:
Actually wasn't being used earlier in hydration code, but added in this commit to ensure that `hydration` is invoked on valid kpt packages. Valid kpt package will have `Kptfile`. I am on the fence if I want to make this `invariant` part of `pkg.New()` constructor.

frankfarzan replied:
Are you thinking about the commands such as `kpt pkg init` where the Kptfile does not exist?

droot replied:
No, my comment was more about checking if `Kptfile` is missing in `pkg.New(...)` constructor.

droot commented:
@frankfarzan addressed most of the comments (will mark addressed one resolved). I have addressed the review comments in a separate commit to make it easier to review https://github.com/GoogleContainerTools/kpt/pull/1483/commits/80c01ce072eac55f15f73688b4fb59dc52883a6d

Shell32-Natsu commented on pkg/test/runner/runner.go:
```diff
@@ -114,7 +118,10 @@ func (r *Runner) runPipeline() error {
-	defer os.RemoveAll(tmpDir)
+	if !r.testCase.Config.Debug {
+		// if debug is true, keep the test directory around for debugging
+		defer os.RemoveAll(tmpDir)
```
It's better to output the path of temporary directory

droot replied:
Good pt. Added.

Shell32-Natsu commented on pkg/test/runner/runner.go:
```diff
@@ -23,8 +23,8 @@ const (
-	CommandFn           string = "fn"
-	CommandPipeline     string = "pipeline"
+	CommandFnEval       string = "run"
+	CommandFnRender     string = "render"
```
Shouldn't this be `CommandPipelineRender`?

droot replied:
These are actually now `fn` subcommands. We no longer have `pipeline run` command.

frankfarzan commented on internal/pkg/pkg.go:
```diff
@@ -77,26 +92,177 @@ func New(path string) (*Pkg, error) {
+func (p *Pkg) readKptfile() (*kptfilev1alpha2.KptFile, error) {
+	kf := &kptfilev1alpha2.KptFile{}
+
+	f, err := os.Open(path.Join(p.UniquePath.String(), kptfilev1alpha2.KptFileName))
```
Do you actually need `UniquePath.String()`? I find think of String() to be mostly relevant for printing purposes not for computation. Why not just `string(p.UniquePath)`.

droot replied:
Changed this one to `string(p.UniquePath)`.

frankfarzan commented on internal/pkg/pkg.go:
```diff
@@ -77,26 +92,177 @@ func New(path string) (*Pkg, error) {
+	}
+	if !includeMetaResources {
+		resources = filterMetaResources(resources)
+		resources, err = filterFunctionConfigFiles(resources, functionConfigFilePaths(pl))
```
Should this be folded into filterMetaResources()?

droot replied:
Done.

frankfarzan commented on internal/pkg/pkg.go:
```diff
@@ -77,26 +92,177 @@ func New(path string) (*Pkg, error) {
+	return resources, err
+}
+
+// filterMetaResources filters kpt metadata files such as Kptfile, function configs.
```
I think we have defined meta resources as including function configs, no?

droot replied:
Yes, done.

frankfarzan reviewed (commented):
Looking good

droot commented on internal/pipeline/executor.go:
```diff
@@ -207,78 +135,107 @@ func (s hydrationState) String() string {
+			return resources, fmt.Errorf("package %s detected in invalid state", currPkg.pkg)
 		}
-		return resources, err
+	} else {
```
Done.

droot commented on internal/pipeline/executor.go:
```diff
@@ -207,78 +135,107 @@ func (s hydrationState) String() string {
-			if depPkg, err = newPkg(s); err != nil {
-				return resources, err
-			}
+	resources, err = currPkg.runPipeline(input, hctx)
```
Using `input` and `output`, reading much better. Thanks.

droot commented on internal/pipeline/executor.go:
```diff
@@ -36,161 +35,90 @@ type Executor struct {
+}
+
+func (pn *pkgNode) Path() string {
+	return string(pn.pkg.UniquePath)
```
I agree. Last time I tried using `UniquePath`, I was running into issues where I had to convert `UniquePath` to string in ton of places because hydration use filesystem package extensively, so I gave up. Now that I have code almost final, I want to try again. I am keeping this comment open.

droot replied:
Made another attempt today to use `pkg.UniquePath` type and worked pretty well this time. Tracked these changes in a separate commit to make easier to review.

droot commented on internal/pipeline/executor.go:
```diff
@@ -207,78 +135,107 @@ func (s hydrationState) String() string {
+		// path annotation in each resource points to path relative to that package.
+		// But the resources are written to the file system at the root package level, so
+		// the path annotation in each resources needs to be adjusted to be relative to the rootPkg.
+		relPath, err := filepath.Rel(hctx.root.Path(), currPkg.Path())
```
Defined pkg level `RelativePathTo` method as you suggested in `1`. Pl. take a look and let me know.

droot commented on internal/pipeline/executor.go:
```diff
@@ -207,78 +135,107 @@ func (s hydrationState) String() string {
+	}
+
+	// empty pipeline
+	if len(pl.Mutators) == 0 && len(pl.Validators) == 0 {
```
Added `pipeline.IsEmpty()` method to encapsulate the logic.

natasha41575 commented on pkg/test/runner/runner.go:
```diff
@@ -45,10 +45,14 @@ func NewRunner(testCase TestCase, c string) (*Runner, error) {
+	case CommandFnEval:
 		return r.runFn()
+	case CommandFnRender:
+		return r.runPipeline()
```
Maybe these method names should change. Before they were calling `kpt fn run` and `kpt pipeline run` and now the names here don't make sense. E.g. instead of `r.runFn` and `r.runPipeline` they can be `r.fnRun` and `r.fnRender`

droot replied:
Good point. Done.

droot commented on internal/pipeline/executor.go:
```diff
@@ -207,78 +135,107 @@ func (s hydrationState) String() string {
+		return input, nil
+	}
+
+	filters, err := fnFilters(pl, pn.Path())
```
Yeah. Removed an unnecessary function and calling it `fnChain`.

droot commented on internal/pipeline/executor.go:
```diff
@@ -288,33 +245,35 @@ func hydrate(p *pkg, hctx *hydrationContext) (resources []*yaml.RNode, err error
+		}
+		// TODO(droot): revisit this. There is an edge case where if relativePath(root, curr)
+		// is same as relativePath(curr, local-resource)
+		if relPath != "" && !strings.HasPrefix(meta.Annotations[kioutil.PathAnnotation], relPath) {
```
Done.

droot commented on internal/pipeline/executor.go:
```diff
@@ -288,33 +245,35 @@ func hydrate(p *pkg, hctx *hydrationContext) (resources []*yaml.RNode, err error
-// Kptfile, permissions etc.
-func filterMetaData(resources []*yaml.RNode) []*yaml.RNode {
-	var filtered []*yaml.RNode
+// adjustRelPath updates the resources with given relative path.
```
Done. Also added new method at the `pkg` level for `RelativePathTo` that takes another package.

droot commented on internal/pipeline/testdata/app-with-generator/configs/db/Kptfile:
```diff
@@ -0,0 +1,20 @@
+  name: db
+pipeline:
+  mutators:
+  - image: gcr.io/sunilarora-sandbox/httpbin-fn:latest
```
I plan to move this function to function catalog under `contrib` section. This will point to the function from the catalog before this PR is merged, so will keep this comment unresolved for now.

droot replied:
Initiated a PR for adding this function in function catalog: https://github.com/GoogleContainerTools/kpt-functions-catalog/pull/182

droot commented on internal/pipeline/testdata/app-with-db/configs/resources.yaml:
```diff
@@ -0,0 +1,13 @@
+apiVersion: apps/v1
```
Ack.

droot commented on internal/pipeline/testdata/simplepackage/configs/Kptfile:
```diff
@@ -4,9 +4,18 @@ metadata:
-    configMap:
-      tier: backend
+  - image: gcr.io/kpt-fn/set-label:unstable
+    config:
```
Done.

droot commented on internal/pkg/pkg.go:
```diff
@@ -37,25 +52,25 @@ type Pkg struct {
-	u, err := filepath.EvalSymlinks(p)
+	// Note(droot): Calling EvalSymlinks before Abs causes a strange issue on OSX
+	// for directories under /tmp. If you have directory /tmp/a,
+	// evalSymlinks("a") !==> /private/tmp
```
Yes.

droot commented on internal/pkg/pkg.go:
```diff
@@ -37,25 +52,25 @@ type Pkg struct {
 	p := filepath.Clean(path)
 
-	u, err := filepath.EvalSymlinks(p)
+	// Note(droot): Calling EvalSymlinks before Abs causes a strange issue on OSX
```
So did some digging:
On my laptop (OSX), `/tmp` and `/var` are symlinks pointing to `/private/tmp` and `/private/var` respectively and luckily, I ran into this issue during a manual test. So if `EvalSymlinks` is called on a path that contains a symlinks component for ex. calling `EvalSymlinks(`/tmp/some-dir`)` will resolve to `/private/tmp/some-dir` and eventually `Abs(`/private/tmp/some-dir`)` resulting in `/private/tmp/some-dir`).  Versus, assuming you are under `/tmp` dir, calling `EvalSymlinks(`some-dir`)` ==> `some-dir` and Abs(`some-dir`) resulting in `/tmp/some-dir`. 

The above scenario is not a problem if you don't have nested package hierarchy, this becomes a problem in a recursive hydration because subpackages will have absolute paths `/tmp/some-dir/subpkg-a` and they will resolve to `/private/tmp/some-dir/subpkg-a` resulting in an inconsistent filesystem hierarchy for the pkg.

So changing the order and calling `Abs(..)` first forces the symlinks to be resolved always (or calling Abs on the rootpkg before calling pkg.New() can avoid the issue).

Now, I expect to see same behavior on Linux as well, but haven't verified it.

droot replied:
Updated the comment with more details.

droot commented on internal/pipeline/executor.go:
```diff
@@ -288,33 +245,35 @@ func hydrate(p *pkg, hctx *hydrationContext) (resources []*yaml.RNode, err error
+		if err != nil {
+			return resources, err
+		}
+		// TODO(droot): revisit this. There is an edge case where if relativePath(root, curr)
```
Worked through an example today and verified that this edge case doesn't exist. resources will be stamped with correct relative path (root --> pkg) when they are either read or generated at that package. Subsequent call to `adjustRelPath` up in the chain (as we move from inward pkg to outer) will always have relative path prefixed. This use-case is exercised using two e2e test cases that we have `pkg-with-generator` and `pkg-with-subpkgs`, so we are good from testing standpoint.

I was initially thinking of using `pkgbuilder` under testutil to implement some more unit tests but that pkg requires more tweaking because it uses older `kptfile` format, so abandoned that idea.

droot commented:
@frankfarzan I believe I have addressed all the feedback. Pl. take a look. One thing I want to improve in a follow up PR:

- Error strings/context for end-users and debugging
- Possibly a way to implement lightweight unit tests which is a bit challenging given that the core logic depends heavily on filesystem calls and function-runner through docker, but I want to make an attempt because that can a long way.

frankfarzan commented on internal/pkg/pkg.go:
```diff
@@ -150,6 +154,15 @@ func (p *Pkg) String() string {
 	return string(p.DisplayPath)
 }
 
+// RelativePathTo returns current package's path relative to a given package.
```
Explain the semantics of the output string. It's OS-specific path (not slash seperated).

frankfarzan reviewed (approved):
Regarding test coverage: In addition to increasing unit test coverage, #1548 will address conversion of function catalog examples.
